akka-remote/src/main/scala/akka/remote/Endpoint.scala

```diff
@@ -312,10 +313,15 @@ private[remote] class ReliableDeliverySupervisor(
       writer forward s
   }
 
-  def gated: Receive = {
-    case Terminated(_) ⇒
+  def gated(writerTerminated: Boolean): Receive = {
+    case Terminated(_) if !writerTerminated ⇒
       context.system.scheduler.scheduleOnce(settings.RetryGateClosedFor, self, Ungate)
+      context.become(gated(writerTerminated = true))
     case IsIdle ⇒ sender() ! Idle
+    case Ungate if !writerTerminated ⇒
+      // Ungate was sent from EndpointManager, but we must wait for Terminated first.
+      // Try again soon, it will be Terminated in just a moment.
+      context.system.scheduler.scheduleOnce(100.millis, self, Ungate)
```

Comment:
we could also store the fact that Ungate was received in another boolean flag and then react accordingly on line 319